### PR TITLE
APB-8296 model encryption

### PR DESCRIPTION
**Pull request by nubz**

### CODE REVIEW

File: app/uk/gov/hmrc/agentsubscription/model/BusinessAddress.scala [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+      for {
+        isEncrypted <- (json \ "encrypted").validateOpt[Boolean]
+        businessAddress = BusinessAddress(
+                            decryptString("addressLine1", isEncrypted, json),
+                            decryptOptString("addressLine2", isEncrypted, json),
+                            decryptOptString("addressLine3", isEncrypted, json),
+                            decryptOptString("addressLine4", isEncrypted, json),
+                            decryptOptString("postalCode", isEncrypted, json),
+                            decryptString("countryCode", isEncrypted, json),
+                            Some(true)
```
Comment (r-melvin) [inline app/uk/gov/hmrc/agentsubscription/model/BusinessAddress.scala:46]:
```suggestion
                            isEncrypted
```

File: app/uk/gov/hmrc/agentsubscription/model/ContactEmailData.scala [hunk 2]
```diff
@@ -16,10 +16,35 @@
...
+
+    def reads(json: JsValue): JsResult[ContactEmailData] =
+      for {
+        isEncrypted <- (json \ "encrypted").validateOpt[Boolean]
+        result = ContactEmailData(
+                   (json \ "useBusinessEmail").as[Boolean],
+                   decryptOptString("contactEmail", isEncrypted, json),
+                   (json \ "encrypted").asOpt[Boolean]
+                 )
+      } yield result
```
Comment (r-melvin) [inline app/uk/gov/hmrc/agentsubscription/model/ContactEmailData.scala:37]:
```suggestion
                  useBusinessEmail <- (json \ "useBusinessEmail").validate[Boolean],
                   contactEmail = decryptOptString("contactEmail", isEncrypted, json),
      } yield ContactEmailData(useBusinessEmail, contactEmail, isEncrypted)
```

File: app/uk/gov/hmrc/agentsubscription/model/ContactTradingAddressData.scala [hunk 3]
```diff
@@ -16,12 +16,35 @@
...
+      val useBusinessAddress = (json \ "useBusinessAddress").as[Boolean]
+      val contactTradingAddress =
+        (json \ "contactTradingAddress").asOpt[BusinessAddress](BusinessAddress.databaseFormat(crypto))
+      JsSuccess(
+        ContactTradingAddressData(
+          useBusinessAddress,
+          contactTradingAddress
+        )
+      )
+    }
```
Comment (r-melvin) [inline app/uk/gov/hmrc/agentsubscription/model/ContactTradingAddressData.scala:37]:
```suggestion
for {
       useBusinessAddress <- (json \ "useBusinessAddress").validate[Boolean]
    contactTradingAddress <- (json \ "contactTradingAddress").validateOpt[BusinessAddress](BusinessAddress.databaseFormat(crypto))
    } yield 
        ContactTradingAddressData(
          useBusinessAddress,
          contactTradingAddress)
```

File: app/uk/gov/hmrc/agentsubscription/model/ContactTradingNameData.scala [hunk 4]
```diff
@@ -16,11 +16,39 @@
...
-  implicit val format: Format[ContactTradingNameData] = Json.format[ContactTradingNameData]
+    def reads(json: JsValue): JsResult[ContactTradingNameData] =
+      for {
+        isEncrypted <- (json \ "encrypted").validateOpt[Boolean]
+        result = ContactTradingNameData(
+                   (json \ "hasTradingName").as[Boolean],
+                   decryptOptString("contactTradingName", isEncrypted, json),
+                   isEncrypted
+                 )
+      } yield result
```
Comment (r-melvin) [inline app/uk/gov/hmrc/agentsubscription/model/ContactTradingNameData.scala:41]:
```suggestion
      for {
        isEncrypted <- (json \ "encrypted").validateOpt[Boolean]
        hasTradingName <- (json \ "hasTradingName")validate[Boolean]
                   contactTradingName = decryptOptString("contactTradingName", isEncrypted, json)
      } yield ContactTradingName(hasTradingName, contactTradingName, isEncrypted)
```

File: app/uk/gov/hmrc/agentsubscription/model/subscriptionJourney/ContactTelephoneData.scala [hunk 5]
```diff
@@ -16,11 +16,39 @@
...
-  implicit val format: Format[ContactTelephoneData] = Json.format[ContactTelephoneData]
+    def reads(json: JsValue): JsResult[ContactTelephoneData] =
+      for {
+        isEncrypted <- (json \ "encrypted").validateOpt[Boolean]
+        result = ContactTelephoneData(
+                   (json \ "useBusinessTelephone").as[Boolean],
+                   decryptOptString("telephoneNumber", isEncrypted, json),
+                   (json \ "encrypted").asOpt[Boolean]
+                 )
+      } yield result
```
Comment (r-melvin) [inline app/uk/gov/hmrc/agentsubscription/model/subscriptionJourney/ContactTelephoneData.scala:41]: Should probably yield the ContactTelephoneData as the other examples have done for a consistent code style

File: app/uk/gov/hmrc/agentsubscription/model/subscriptionJourney/Registration.scala [hunk 6]
```diff
@@ -26,11 +29,44 @@ case class Registration(
...
+                   decryptOptString("taxpayerName", isEncrypted, json),
+                   (json \ "isSubscribedToAgentServices").as[Boolean],
+                   (json \ "isSubscribedToETMP").as[Boolean],
+                   (json \ "address").as[BusinessAddress](BusinessAddress.databaseFormat(crypto)),
+                   decryptOptString("emailAddress", isEncrypted, json),
+                   decryptOptString("primaryPhoneNumber", isEncrypted, json),
+                   (json \ "safeId").asOpt[String],
+                   isEncrypted
+                 )
+      } yield result
```
Comment (r-melvin) [inline app/uk/gov/hmrc/agentsubscription/model/subscriptionJourney/Registration.scala:53]: as above

File: app/uk/gov/hmrc/agentsubscription/model/BusinessAddress.scala (hunk 1 above)
Comment (nubz) [reply app/uk/gov/hmrc/agentsubscription/model/BusinessAddress.scala]: done

File: app/uk/gov/hmrc/agentsubscription/model/ContactEmailData.scala (hunk 2 above)
Comment (nubz) [reply app/uk/gov/hmrc/agentsubscription/model/ContactEmailData.scala]: done

File: app/uk/gov/hmrc/agentsubscription/model/ContactTradingAddressData.scala (hunk 3 above)
Comment (nubz) [reply app/uk/gov/hmrc/agentsubscription/model/ContactTradingAddressData.scala]: done

File: app/uk/gov/hmrc/agentsubscription/model/ContactTradingNameData.scala (hunk 4 above)
Comment (nubz) [reply app/uk/gov/hmrc/agentsubscription/model/ContactTradingNameData.scala]: done

File: app/uk/gov/hmrc/agentsubscription/model/subscriptionJourney/ContactTelephoneData.scala (hunk 5 above)
Comment (nubz) [reply app/uk/gov/hmrc/agentsubscription/model/subscriptionJourney/ContactTelephoneData.scala]: done

File: app/uk/gov/hmrc/agentsubscription/model/subscriptionJourney/Registration.scala (hunk 6 above)
Comment (nubz) [reply app/uk/gov/hmrc/agentsubscription/model/subscriptionJourney/Registration.scala]: done

**Review by geoffreywatson — APPROVED** [review]

**Review by geoffreywatson — APPROVED** [review]

**Pull request merged**